msm/src/serialization/witness.rs

```diff
@@ -175,62 +175,45 @@ impl<Fp: PrimeField> Env<Fp> {
         }
     }
 
-    pub fn add_rangecheck4_table_value(&mut self, i: usize) {
-        if i < (1 << 4) {
-            self.lookup_t_multiplicities_rangecheck4[i] += Fp::one();
-        } else {
-            self.lookup_t_multiplicities_rangecheck4[0] += Fp::one();
-        }
-    }
-
     pub fn reset(&mut self) {
-        self.rangecheck15_lookups = vec![];
-        self.rangecheck4_lookups = vec![];
+        *self.lookups.get_mut(&LookupTable::RangeCheck4).unwrap() = Vec::new();
+        *self.lookups.get_mut(&LookupTable::RangeCheck15).unwrap() = Vec::new();
     }
 
-    /// Return the normalized multiplicity vector of RangeCheck4 in case the
-    /// table is not injective. Note that it is the case for `RangeCheck4`.
-    pub fn get_rangecheck4_normalized_multipliticies(
-        &self,
-        domain: EvaluationDomains<Fp>,
-    ) -> Vec<Fp> {
-        let mut m = vec![Fp::zero(); 1 << 4];
-        self.lookup_multiplicities_rangecheck4
-            .into_iter()
-            .zip(self.lookup_t_multiplicities_rangecheck4.iter())
-            .enumerate()
-            .for_each(|(i, (m_f, m_t))| m[i] = m_f / m_t);
-        let repeated_dummy_value: Vec<Fp> = iter::repeat(m[0])
+    pub fn get_rangecheck4_multipliticies(&self, domain: EvaluationDomains<Fp>) -> Vec<Fp> {
+        let mut m = Vec::with_capacity(domain.d1.size as usize);
+        m.extend(self.lookup_multiplicities[&LookupTable::RangeCheck4].to_vec());
+        let repeated_dummy_value: Vec<Fp> = iter::repeat(-Fp::zero())
```

Comment:
I see, so it is indeed a dummy repeated value, in this case the first entry which turns out to be zero.